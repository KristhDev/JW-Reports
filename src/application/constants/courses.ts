export const coursesMessages = {
    EXPORTED_SUCCESS: 'Has exportado tus cursos correctamente. El archivo se encuentra en su carpeta de descargas.',
    PERSON_NAME_REQUIRED: 'El nombre de la persona es requerido.',
    PERSON_NAME_MIN_LENGTH: 'El nombre de la persona debe tener al menos 2 caracteres.',
    PERSON_ABOUT_REQUIRED: 'La información de la persona es requerida.',
    PERSON_ABOUT_MIN_LENGTH: 'La información de la persona debe tener al menos 10 caracteres.',
    PERSON_ADDRESS_REQUIRED: 'La dirección es requerida.',
    PERSON_ADDRESS_MIN_LENGTH: 'La dirección debe tener al menos 10 caracteres.',
    PUBLICATION_REQUIRED: 'La publicación de estudio no puede estar vacía.',
    PUBLICATION_MIN_LENGTH: 'La publicación de estudio debe tener al menos 5 caracteres.',
    UNSELECTED: 'No hay un curso seleccionado.',
    UNSELECTED_FINISH_OR_START: 'No pudes terminar o comenzar de nuevo un curso suspendido.',
    UNSELECTED_DELETE: 'No hay un curso seleccionado para eliminar.',
    UNSELECTED_UPDATE: 'No hay un curso seleccionado para actualizar.',
    FINISHED: 'No puedes suspender o renovar un curso terminado.',
    SUSPENDED_SUCCESS: 'Has suspendido el curso correctamente.',
    RENEW_SUCCESS: 'Has renovado el curso correctamente.',
    DELETED_SUCCESS: 'Has eliminado el curso correctamente.',
    FINISHED_SUCCESS: 'Has terminado el curso correctamente.',
    RESTARTED_SUCCESS: 'Has comenzado de nuevo el curso correctamente.',
    ADDED_SUCCESS: 'Has agregado el curso correctamente.',
    UPDATED_SUCCESS: 'Has actualizado el curso correctamente.',
}