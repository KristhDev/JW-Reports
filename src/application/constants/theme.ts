import { ItemOption } from '../../infrasturcture/interfaces';

/**
 * The theme options to select
 */
export const THEME_OPTIONS: ItemOption[] = [
    { label: 'Modo predeterminado', value: 'default' },
    { label: 'Modo claro', value: 'light' },
    { label: 'Modo oscuro', value: 'dark' }
];