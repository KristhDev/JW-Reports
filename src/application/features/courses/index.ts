export {
    default as coursesReducer,
    INIT_COURSE,
    COURSES_INITIAL_STATE,
    addCourse,
    addCourses,
    addLastLessonInCourse,
    clearCourses,
    removeCourse,
    removeCourses,
    replaceLastLessonInCourse,
    setCourseFilter,
    setCourses,
    setCoursesPagination,
    setCoursesScreenHistory,
    setHasMoreCourses,
    setIsCourseDeleting,
    setIsCourseLoading,
    setIsCoursesExporting,
    setIsCoursesLoading,
    setRefreshCourses,
    setSelectedCourse,
    updateCourse,
    updateLastLessonInCourse
} from './slice';

export * from './types';