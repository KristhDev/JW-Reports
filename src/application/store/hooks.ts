import { useDispatch, useSelector } from 'react-redux';

/* Store */
import { AppDispatch, RootState } from './store';

/* Defined types hooks */
export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export const useAppSelector = useSelector.withTypes<RootState>();