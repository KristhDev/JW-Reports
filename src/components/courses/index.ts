export { CourseCard } from './CourseCard';
export { CoursesList } from './CoursesList';