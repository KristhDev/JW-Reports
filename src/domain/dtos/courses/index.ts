export * from './active-suspend.dto';
export * from './create.dto';
export * from './finish-start.dto';
export * from './update.dto';