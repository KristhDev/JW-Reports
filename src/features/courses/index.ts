export {
    default as coursesReducer,
    INIT_COURSE,
    INIT_LESSON,
    addCourse,
    addCourses,
    addLesson,
    addLessons,
    clearCourses,
    removeCourse,
    removeCourses,
    removeLesson,
    removeLessons,
    setCourseFilter,
    setCourses,
    setCoursesPagination,
    setCoursesScreenHistory,
    setHasMoreCourses,
    setHasMoreLessons,
    setIsCourseDeleting,
    setIsCourseLoading,
    setIsCoursesLoading,
    setIsLessonDeleting,
    setIsLessonLoading,
    setIsLessonsLoading,
    setLessons,
    setLessonsPagination,
    setRefreshCourses,
    setRefreshLessons,
    setSelectedCourse,
    setSelectedLesson,
    updateCourse,
    updateLesson
} from './slice';