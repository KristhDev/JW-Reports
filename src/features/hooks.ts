import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';

/* Store */
import { AppDispatch, RootState } from './store';

/* Defined types hooks */
export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;