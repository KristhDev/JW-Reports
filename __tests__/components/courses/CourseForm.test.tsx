import React from 'react';
import { act, fireEvent, render, screen, waitFor } from '@testing-library/react-native';

/* Components */
import { CourseForm } from '../../../src/components/courses';

/* Hooks */
import { useCourses, useStatus, useTheme } from '../../../src/hooks';

/* Theme */
import { darkColors } from '../../../src/theme';

/* Mocks */
import { courseSelectedStateMock, coursesStateMock, saveCourseMock, setErrorFormMock, updateCourseMock } from '../../mocks';

/* Mock hooks */
jest.mock('../../../src/hooks/useCourses.ts');
jest.mock('../../../src/hooks/useStatus.ts');
jest.mock('../../../src/hooks/useTheme.ts');

describe('Test in <CourseForm /> component', () => {
    (useCourses as jest.Mock).mockReturnValue({
        state: coursesStateMock,
        saveCourse: saveCourseMock,
        updateCourse: updateCourseMock
    });

    (useStatus as jest.Mock).mockReturnValue({
        setErrorForm: setErrorFormMock
    });

    (useTheme as jest.Mock).mockReturnValue({
        state: { colors: darkColors },
        BUTTON_TRANSPARENT_COLOR: 'rgba(255, 255, 255, 0.15)'
    });

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should to match snapshot', async () => {
        await waitFor(() => {
            render(<CourseForm />);
        });

        await act(() => {
            expect(screen.toJSON()).toMatchSnapshot();
        });
    });

    it('should call setErrorForm when the form is empty or invalid', async () => {
        await waitFor(() => {
            render(<CourseForm />);
        });

        const touchable = await screen.findByTestId('button-touchable');

        await waitFor(() => {
            /* Get submit touchable */
            fireEvent.press(touchable);
        });

        /* Check if setErrorForm is called one time */
        expect(setErrorFormMock).toHaveBeenCalledTimes(1);
    });

    it('should call saveCourse when the form is valid and selectedCourse is empty', async () => {
        await waitFor(() => {
            render(<CourseForm />);
        });

        const personName = 'Virginie';
        const personAbout = 'Fuga sint rerum voluptatem et quaerat consequatur quas iure.';
        const personAddress = 'Commodi dolorem nesciunt vel sunt quo magnam et enim accusamus. Enim error dolorem voluptas et sed accusamus similique quod.';
        const publication = 'eius velit veritatis';

        const inputsText = await  screen.findAllByTestId('form-field-text-input');

        await waitFor(() => {

            /* Get all text inputs to type person_name, person_about, person_address and publication */
            fireEvent(inputsText[0], 'onChangeText', personName);
            fireEvent(inputsText[1], 'onChangeText', personAbout);
            fireEvent(inputsText[2], 'onChangeText', personAddress);
            fireEvent(inputsText[3], 'onChangeText', publication);
        });

        /* Get submit touchable */
        const touchable = await screen.findByTestId('button-touchable');

        await waitFor(() => {
            fireEvent.press(touchable);
        });

        /* Check if setErrorForm is called one time */
        expect(saveCourseMock).toHaveBeenCalledTimes(1);

        /* Get text of submit touchable */
        const btnText = await screen.findByTestId('button-text');

        /**
         * Check if text of submit touchable is equal to Guardar and saveCourse
         * is called with respective args
         */
        expect(btnText.props.children).toBe('Guardar');
        expect(saveCourseMock).toHaveBeenCalledWith({ personName, personAbout, personAddress, publication });
    });

    it('should call updateCourse when the form is valid and selectedCourse isnt empty', async () => {

        /* Mock data of useCourses */
        (useCourses as jest.Mock).mockReturnValue({
            state: courseSelectedStateMock,
            saveCourse: saveCourseMock,
            updateCourse: updateCourseMock
        });

        await waitFor(() => {
            render(<CourseForm />);
        });

        const personAbout = 'Aliquid dolore quae et voluptatibus dicta.';
        const inputsText = await screen.findAllByTestId('form-field-text-input');

        await waitFor(() => {

            /* Get text input of about to type new content */
            fireEvent(inputsText[1], 'onChangeText', personAbout);
        });

        /* Get submit touchable */
        const touchable = await screen.findByTestId('button-touchable');

        await waitFor(() => {
            fireEvent.press(touchable);
        });

        /* Check if updateCourseMock is called one time */
        expect(updateCourseMock).toHaveBeenCalledTimes(1);

        /* Get text of submit touchable */
        const btnText = await screen.findByTestId('button-text');

        /**
         * Check if text of submit touchable is equal to Actualizar and updateCourse
         * is called with respective args
         */
        expect(btnText.props.children).toBe('Actualizar');
        expect(updateCourseMock).toHaveBeenCalledWith({
            personName: courseSelectedStateMock.selectedCourse.personName,
            personAbout: personAbout,
            personAddress: courseSelectedStateMock.selectedCourse.personAddress,
            publication: courseSelectedStateMock.selectedCourse.publication
        });
    });
});