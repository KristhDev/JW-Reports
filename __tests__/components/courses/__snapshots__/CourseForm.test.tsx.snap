// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <CourseForm /> component should to match snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "paddingBottom": 40,
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "display": "flex",
        "marginBottom": 16,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#FFFFFF",
            "fontSize": 18,
            "marginBottom": 16,
          },
          undefined,
        ]
      }
      testID="form-field-label"
    >
      Nombre del estudiante:
    </Text>
    <View
      style={
        {
          "borderColor": "transparent",
          "borderRadius": 7,
          "borderWidth": 1,
        }
      }
    >
      <View
        style={
          {
            "borderColor": "#FFFFFF",
            "borderRadius": 5,
            "borderWidth": 0.5,
          }
        }
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "borderColor": "transparent",
                "borderRadius": 5,
                "borderWidth": 1.5,
                "flexDirection": "row",
                "paddingRight": 10,
              },
              undefined,
            ]
          }
        >
          <TextInput
            autoCorrect={false}
            cursorColor="#C0A7E1"
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onSelectionChange={[Function]}
            placeholder="Ingrese el nombre"
            placeholderTextColor="#BFBFBF"
            selection={
              {
                "end": 0,
                "start": 0,
              }
            }
            selectionColor="#A0B9E2"
            style={
              [
                {
                  "color": "#FFFFFF",
                  "flex": 1,
                  "fontSize": 17,
                  "paddingLeft": 10,
                  "paddingRight": 5,
                  "textAlignVertical": "center",
                },
                undefined,
              ]
            }
            testID="form-field-text-input"
            value=""
          />
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "#BFBFBF",
                  "fontSize": 25,
                },
                undefined,
                {
                  "fontFamily": "Ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "display": "flex",
        "marginBottom": 16,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#FFFFFF",
            "fontSize": 18,
            "marginBottom": 16,
          },
          undefined,
        ]
      }
      testID="form-field-label"
    >
      Información del estudiante:
    </Text>
    <View
      style={
        {
          "borderColor": "transparent",
          "borderRadius": 7,
          "borderWidth": 1,
        }
      }
    >
      <View
        style={
          {
            "borderColor": "#FFFFFF",
            "borderRadius": 5,
            "borderWidth": 0.5,
          }
        }
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "borderColor": "transparent",
                "borderRadius": 5,
                "borderWidth": 1.5,
                "flexDirection": "row",
                "paddingRight": 10,
              },
              undefined,
            ]
          }
        >
          <TextInput
            autoCorrect={false}
            cursorColor="#C0A7E1"
            multiline={true}
            numberOfLines={10}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onSelectionChange={[Function]}
            placeholder="Ingrese datos sobre la persona, temas de interés, preferencias, aspectos importantes, etc..."
            placeholderTextColor="#BFBFBF"
            selection={
              {
                "end": 0,
                "start": 0,
              }
            }
            selectionColor="#A0B9E2"
            style={
              [
                {
                  "color": "#FFFFFF",
                  "flex": 1,
                  "fontSize": 17,
                  "paddingLeft": 10,
                  "paddingRight": 5,
                  "textAlignVertical": "top",
                },
                undefined,
              ]
            }
            testID="form-field-text-input"
            value=""
          />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "display": "flex",
        "marginBottom": 16,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#FFFFFF",
            "fontSize": 18,
            "marginBottom": 16,
          },
          undefined,
        ]
      }
      testID="form-field-label"
    >
      Dirección:
    </Text>
    <View
      style={
        {
          "borderColor": "transparent",
          "borderRadius": 7,
          "borderWidth": 1,
        }
      }
    >
      <View
        style={
          {
            "borderColor": "#FFFFFF",
            "borderRadius": 5,
            "borderWidth": 0.5,
          }
        }
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "borderColor": "transparent",
                "borderRadius": 5,
                "borderWidth": 1.5,
                "flexDirection": "row",
                "paddingRight": 10,
              },
              undefined,
            ]
          }
        >
          <TextInput
            autoCorrect={false}
            cursorColor="#C0A7E1"
            multiline={true}
            numberOfLines={4}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onSelectionChange={[Function]}
            placeholder="Ingrese la dirección"
            placeholderTextColor="#BFBFBF"
            selection={
              {
                "end": 0,
                "start": 0,
              }
            }
            selectionColor="#A0B9E2"
            style={
              [
                {
                  "color": "#FFFFFF",
                  "flex": 1,
                  "fontSize": 17,
                  "paddingLeft": 10,
                  "paddingRight": 5,
                  "textAlignVertical": "top",
                },
                undefined,
              ]
            }
            testID="form-field-text-input"
            value=""
          />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "display": "flex",
        "marginBottom": 40,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#FFFFFF",
            "fontSize": 18,
            "marginBottom": 16,
          },
          undefined,
        ]
      }
      testID="form-field-label"
    >
      Publicación de estudio:
    </Text>
    <View
      style={
        {
          "borderColor": "transparent",
          "borderRadius": 7,
          "borderWidth": 1,
        }
      }
    >
      <View
        style={
          {
            "borderColor": "#FFFFFF",
            "borderRadius": 5,
            "borderWidth": 0.5,
          }
        }
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "borderColor": "transparent",
                "borderRadius": 5,
                "borderWidth": 1.5,
                "flexDirection": "row",
                "paddingRight": 10,
              },
              undefined,
            ]
          }
        >
          <TextInput
            autoCorrect={false}
            cursorColor="#C0A7E1"
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onSelectionChange={[Function]}
            placeholder="Ingrese la publicación"
            placeholderTextColor="#BFBFBF"
            selection={
              {
                "end": 0,
                "start": 0,
              }
            }
            selectionColor="#A0B9E2"
            style={
              [
                {
                  "color": "#FFFFFF",
                  "flex": 1,
                  "fontSize": 17,
                  "paddingLeft": 10,
                  "paddingRight": 5,
                  "textAlignVertical": "center",
                },
                undefined,
              ]
            }
            testID="form-field-text-input"
            value=""
          />
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "#BFBFBF",
                  "fontSize": 25,
                },
                undefined,
                {
                  "fontFamily": "Ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "overflow": "hidden",
        },
        [
          {
            "backgroundColor": "#C0A7E1",
            "borderRadius": 10,
          },
          undefined,
        ],
      ]
    }
    testID="button-touchable"
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
            "minWidth": "100%",
            "paddingHorizontal": 20,
            "paddingVertical": 9,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "fontFamily": "System",
              "fontWeight": "400",
              "letterSpacing": 0,
            },
            {
              "writingDirection": "ltr",
            },
            [
              {
                "color": "#121212",
                "fontSize": 18,
                "fontWeight": "bold",
                "marginBottom": 2,
                "textAlign": "center",
              },
              undefined,
            ],
          ]
        }
        testID="button-text"
      >
        Guardar
      </Text>
    </View>
  </View>
</View>
`;
